File: python/kmeans.py
from collections import defaultdict
from math import sqrt
from time import time
import json


class Point(object):
    x = None
    y = None
    def __init__(self, x,y):
        self.x = x
        self.y = y
    def x(self):
        return self.x
    def y(self):
        return self.y
    def __add__(self,other):
        return Point(self.x + other.x, self.y + other.y)
    def __div__(self,value):
        return Point(self.x/float(value),self.y/float(value))
    def dist(self,other):
        return sqrt((self.x - other.x)**2 + (self.y - other.y)**2)
    def closest(self,points):
        return min(points,key=self.dist)
    def __repr__(self):
        return "(%f,%f)" % (self.x,self.y)

def update_centroids(points,centroids):
    groups = groupby(points,centroids)
    res = []
    for g in groups.values():
        res.append(sum(g, Point(0,0))/len(g))
    return res

def groupby(points,centroids):
    g = defaultdict(list)
    for p in points:
        c = p.closest(centroids)
        g[c].append(p)
    return g

def run(xs, n, iters=15):
    centroids = xs[:n]
    for i in xrange(iters):
        centroids = update_centroids(xs,centroids)
    return groupby(xs,centroids)

if __name__ == "__main__":
    with open("../points.json") as f:
        points = map(lambda x: Point(x[0],x[1]),json.loads(f.read()))
    iterations = 100
    start = time()
    for i in xrange(iterations):
        run(points, 10)
    total = (time() - start) * 1000 / iterations
    print("Made {:d} iterations with an average of {:.2f} milliseconds".format(iterations, total))

File: lua/kmeans.lua
local sqrt = math.sqrt
local clock = os.clock
local json = require("dkjson")

local Point = {}
Point.__index = Point

setmetatable(Point, {
	__call = function(self, x, y)
		return setmetatable({_x=x, _y=y}, Point)
	end
})

function Point:x()
	return self._x
end
function Point:y()
	return self._y
end
function Point:__add(other)
	return Point(self._x + other._x, self._y + other._y)
end
function Point:__div(value)
	return Point(self._x / value, self._y / value)
end
function Point:dist(other)
	return sqrt((self._x - other._x)^2 + (self._y - other._y)^2)
end
function Point:closest(points)
	local min = math.huge
	local min_i
	for i, point in ipairs(points) do
		local dist = self:dist(point)
		if dist < min then
			min = dist
			min_i = i
		end
	end
	return points[min_i]
end
function Point:__tostring()
	return ("Point(%f,%f)"):format(self._x, self._y)
end

local function sum(t, start)
	local total = start or 0
	for _, i in ipairs(t) do
		total = total + i
	end
	return total
end

local function len(t)
	local l = 0
	for _ in pairs(t) do
		l = l + 1
	end
	return l
end

local function groupby(points, centroids)
	local g = {}
	for _, p in ipairs(points) do
		local c = p:closest(centroids)
		g[c] = g[c] or {}
		table.insert(g[c], p)
	end
	return g
end

local function update_centroids(points, centroids)
	local groups = groupby(points, centroids)
	local res = {}
	local i = 1
	for _, g in pairs(groups) do
		res[i] = sum(g, Point(0, 0)) / len(g)
		i = i + 1
	end
	return res
end

local function run(xs, n, iters)
	iters = iters or 15
	local centroids = {}
	for i = 1, n do
		table.insert(centroids, xs[i])
	end
	for i = 1, iters do
		centroids = update_centroids(xs, centroids)
	end
	return groupby(xs, centroids)
end

if not ... then
	local data = json.decode(io.open("../points.json"):read("*a"))
	local points = {}
	for i, x in ipairs(data) do
		points[i] = Point(x[1], x[2])
	end
	local start = clock()
	local iterations = 100
	for i = 1, iterations do
		run(points, 10)
	end
	local total = (clock() - start) * 1000 / iterations
	print(("Made %d iterations with an average of %.2f milliseconds"):format(iterations, total))
end
